src/execution_plan/ops/op_update.c

```diff
@@ -18,122 +19,135 @@ static OpResult UpdateReset(OpBase *opBase);
+static int _UpdateEdge
+(
+	OpUpdate *op,
+	PendingUpdateCtx *updates,
+	uint update_count
+) {
+	// Retrieve GraphEntity:
+	// Due to Record freeing we can't maintain the original pointer to
+	// GraphEntity object, but only a pointer to an Entity object, to use the
+	// GraphEntity_Get, GraphEntity_Add functions we'll use a place holder to
+	// hold our entity
+	int attributes_set = 0;
+	GraphEntity *ge = (GraphEntity *)&updates->e;
+
+	for(uint i = 0; i < update_count; i++) {
+		PendingUpdateCtx *update = updates + i;
+		attributes_set += _UpdateEntity(ge, update);
+	}
 
-	int label_id = Graph_GetEdgeRelation(op->gc->g, edge);
+	return attributes_set;
+}
 
-	// Try to get current property value.
-	SIValue *old_value = GraphEntity_GetProperty((GraphEntity *)edge, ctx->attr_id);
+// set a property on an edge. For non-NULL values, the property will be added
+// or updated if it is already present For NULL values, the property will be
+// deleted if present and nothing will be done otherwise Relevant indexes will
+// be updated if required Returns 1 if a property was set or deleted
+static int _UpdateNode
+(
+	OpUpdate *op,
+	PendingUpdateCtx *updates,
+	uint update_count
+) {
+	// retrieve GraphEntity:
+	// due to Record freeing we can't maintain the original pointer to
+	// GraphEntity object, but only a pointer to an Entity object, to use the
+	// GraphEntity_Get, GraphEntity_Add functions we'll use a place holder to
+	// hold our entity
+	int          attributes_set  =  0;
+	bool         update_index    =  false;
+	Node         *node           =  &updates->n;
+	GraphEntity  *ge             =  (GraphEntity  *)node;
+
+	for(uint i = 0; i < update_count; i++) {
+		PendingUpdateCtx  *update    =  updates + i;
+		attributes_set += _UpdateEntity(ge, update);
+		// do we need to update an index for this property?
+		update_index |= update->update_index;
+	}
 
-	if(old_value == PROPERTY_NOTFOUND) {
-		// Adding a new property; do nothing if its value is NULL.
-		if(SI_TYPE(ctx->new_value) == T_NULL) return 0;
-		GraphEntity_AddProperty((GraphEntity *)edge, ctx->attr_id, ctx->new_value);
-	} else {
-		// Update property.
-		GraphEntity_SetProperty((GraphEntity *)edge, ctx->attr_id, ctx->new_value);
+	// update index for node entities if indexed fields have been modified
+	if(update_index) {
+		int label_id = node->labelID;
+		Schema *s = GraphContext_GetSchemaByID(op->gc, label_id, SCHEMA_NODE);
+		_UpdateIndex(node, s);
 	}
-	return 1;
+
+	return attributes_set;
 }
 
-/* Executes delayed updates. */
-static void _CommitUpdates(OpUpdate *op) {
-	uint properties_set = 0;
-	for(uint i = 0; i < op->pending_updates_count; i++) {
-		EntityUpdateCtx *ctx = &op->pending_updates[i];
-		if(ctx->entity_type == GETYPE_NODE) {
-			properties_set += _UpdateNode(op, ctx);
+static void _CommitEntityUpdates(OpUpdate *op, EntityUpdateCtx *ctx) {
+	uint  properties_set      =  0;
+	uint  updates_per_entity  =  array_len(ctx->exps);
```

Comment:
add a comment `total_updates = n * updates_per_entity, where n is the number of entities to update`